service/dap/server.go

```diff
@@ -469,6 +469,123 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
 		return
 	}
 
+	if mode == "record" {
+		s.config.Debugger.Backend = "rr"
+		s.send(&dap.CapabilitiesEvent{ Event: *newEvent("capabilities"), Body: dap.CapabilitiesEventBody{Capabilities: dap.Capabilities{ SupportsStepBack: true }}})
+
+		// Validate rr availability
+		if err := gdbserial.CheckRRAvailable(); err != nil {
+			s.sendErrorResponseWithURL(request.Request, FailedToInitialize, "Unable to find  rr in record configuration", err.Error(), "https://github.com/rr-debugger/rr/wiki/Building-And-Installing", "rr installation page")
+			return
+		}
+
+		// Record only works on rr trace directories
+		traceDirectory, ok := request.Arguments["traceDirectory"].(string)
+		if !ok || traceDirectory == "" {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to launch",
+				"The traceDirectory attribute is missing in record configuration.")
+			return
+		}
+
+		buildFlags := ""
+		buildFlagsArg, ok := request.Arguments["buildFlags"]
+		if ok {
+			buildFlags, ok = buildFlagsArg.(string)
+			if !ok {
+				s.sendErrorResponse(request.Request,
+					FailedToLaunch, "Failed to launch",
+					fmt.Sprintf("'buildFlags' attribute " +
+						"%v' in record configuration is not a string.", buildFlagsArg))
+				return
+			}
+		}
+
+		output, ok := request.Arguments["output"].(string)
+		if !ok || output == "" {
+			output = defaultDebugBinary
+		}
+
+		debugname, err := filepath.Abs(output)
+		err = gobuild.GoBuild(debugname, []string{program}, buildFlags)
+		if err != nil {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to build recording target",
+				fmt.Sprintf("Build error: %s", err.Error()))
+			return
+		}
+
+		traceDir, err := gdbserial.Record([]string{program}, filepath.Dir(program), true, s.config.Debugger.Redirects)
+		if err != nil {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to record trace from target",
+				fmt.Sprintf("Recording error: %s", err.Error()))
+			return
+		}
+		if traceDir == "" {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to record trace from target",
+				"No trace result found")
+			return
+		}
+	}
```

Comment:
I think this code, as it is, is working by accident. You are calling `gdbserial.Record` but throwing away the result, but because of the other work you are doing the call to `debugger.New` later on will re-record the program and replay it. I think everyting inside this `if` can be removed and replaced by something along the lines of:

```
if mode == "record" {
    mode = "debug"
    s.config.Debugger.Backend = "rr"
}
```